### DISCUSSION TITLE
DEBUG Mode

### DISCUSSION THREAD

**Discussion by asadthecoder** [python-smpplib/python-smpplib#157] [Q&A] [answered]
I need your help with following issues. 
1. First how can I disable/enable DEBUG output to stdout and how to redirect to a log?
2. For responses received from ESME e.g. against sm_submit, an PDU is received but can not read message on stdout. Please find below output

DEBUG: Sending submit_sm PDU##########################################################################################################                                 |
DEBUG: >> 000000ad00000004000000000000000b0005005054434c0000013932353134383330303333000000000000010000007d44656172205369722f4d6164616d2c2053616c61727920666f7220746865206d6f6e7468206f662053657074656d626572203230313520686173206265656e20637265646974656420746f20796f75722072657370656374697665206163636f756e742e20526567617264732c204852204f7065726174696f6e732020 173 bytes
DEBUG: Waiting for PDU...
DEBUG: << 0000001980000004000000000000000b363036453930363000 25 bytes
DEBUG: Read submit_sm_resp PDU

submit_sm_resp PDU: How do I read it ?

Thanks

**Comment by code-of-kpp** [comment]
1. We use standard python `logging` module, so by default debug output is just ignored. In `README` we use `logging.basicConfig`  to redirect this output to console.
2. I didn't get your question. `submit_sm_resp` PDU is parsed in `SubmitSMResp` in `command` module.

**Comment by asadthecoder** [comment]
Thank you for your response; and excuse me for not making myself very clear.
1. logging.basicConfig worked just great :) (using INFO for minimal logging)
2. In debug mode if I wanted to parse what really is inside PDU (e.g. sm_submit_resp), I can see the hex characters. Just want to get payload in ascii so I could read it on stdout. For now I have to copy the PDU (in hex) and use some online decoders to read. My ultimate goal is to get delivery report of a every message I send to SMSC for that I need to readout response PDU.

**Comment by code-of-kpp** [comment]
use set_message_sent_handler in client. handler will be called with parsed pdu instance

**Comment by asadthecoder** [comment]
Please excuse me for my very little knowledge but I am very new to Python
and learning more about it.
Could you please guide me how to enable handler and then how to get
delivery report?

Thanks a lot for your great help

On Sun, Sep 20, 2015 at 4:36 PM, Konstantin Podshumok <
notifications@github.com> wrote:

> use set_message_sent_handler in client. handler will be called with parsed
> pdu instance
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/podshumok/python-smpplib/issues/15#issuecomment-141779899
> .

## 

---

Best Wishes

Asad Mehmood

Contact: +92-346-5746693

---

**Comment by code-of-kpp** [comment, marked as answer]
these lines in README set handlers

``` python
client.set_message_sent_handler(
    lambda pdu: sys.stdout.write('sent {} {}\n'.format(pdu.sequence, pdu.message_id)))
client.set_message_received_handler(
    lambda pdu: sys.stdout.write('delivered {}\n'.format(pdu.receipted_message_id)))
```

handler is a function (lambda's here). Client call the handlers, when corresponding PDU is received. `message_sent_handler` is for `submit_sm_resp` and `message_received_handler` is for `deliever_sm`

**Comment by ghost** [comment]
Hello!

I have a same problem, I receive delivered none, I guess it is by the following code:
 `lambda pdu: sys.stdout.write('delivered {}\n'.format(pdu.receipted_message_id))`
and in debug mode I am seeing my data in hex format. 
Please help me why I am getting delivered none.
Tnx
